File: fb/205.Isomorphic-Strings.cpp
// Time O(n)
// Space O(n)
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if(s.size() != t.size()) return false;
        int N = s.size();
        unordered_map<char, char> um;
        for(int i = 0; i < N; i++) {
            if(!um.count(s[i]))
                um[s[i]] = t[i];
            else {
                if(um[s[i]] != t[i])
                    return false;
            }
        }
        um.clear();
        for(int i = 0; i < N; i++) {
            if(!um.count(t[i]))
                um[t[i]] = s[i];
            else {
                if(um[t[i]] != s[i])
                    return false;
            }
        }
        return true;
    }
};

File: fb/205.Isomorphic-Strings.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        map_s_t = {}
        map_t_s = {}
        for c1, c2 in zip(s, t):
            if c1 not in map_s_t and c2 not in map_t_s:
                map_s_t[c1] = c2
                map_t_s[c2] = c1
            elif map_s_t.get(c1) != c2 or map_t_s.get(c2) != c1:
                return False
        return True
